rosbag2_py/test/test_sequential_reader.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pathlib import Path
+
+from common import get_rosbag_options
+from rcl_interfaces.msg import Log
+from rclpy.serialization import deserialize_message
+import rosbag2_py
+from rosidl_runtime_py.utilities import get_message
+from std_msgs.msg import String
+
+
+def test_sequential_reader():
+    bag_path = str(Path(__file__).parent.parent / 'resources' / 'talker')
+    storage_options, converter_options = get_rosbag_options(bag_path)
+
+    reader = rosbag2_py.SequentialReader()
+    reader.open(storage_options, converter_options)
+
+    topic_types = reader.get_all_topics_and_types()
+
+    # Create a map for quicker lookup
+    type_map = {topic_types[i].name: topic_types[i].type for i in range(len(topic_types))}
+
+    # Set filter for topic of string type
+    storage_filter = rosbag2_py.StorageFilter()
+    storage_filter.topics = ['/topic']
+    reader.set_filter(storage_filter)
+
+    msg_counter = 0
+
+    while reader.has_next():
```

Comment:
When doing exactly this I'm getting the following error:

``
RuntimeError: Error when preparing SQL statement 'SELECT data, timestamp, topics.name FROM messages JOIN topics ON messages.topic_id = topics.id ORDER BY messages.timestamp;'. SQLite error: (1): SQL logic error
``

Do you have any idea what could be causing this?

As this code is not yet mainline, I'm not sure where to put the question, happy to create a follow-up issue.

Reply:
The error sounds unrelated to this PR, since it is just adding a wrapper around the C++ API. It could be an issue accessing the DB. Are you able to share code and steps that reproduces the error?

Reply:
The CI tests are passing on all OSes on this PR, so the error shouldn't happen. Do you have the latest version of `rosbag2`? Updating and recompiling everything `--packages-select-regex rosbag2* or just `--packages-select rosbag2` might fix it.

Reply:
The issue is caused by a version 1 rosbag2. Here in the tests a version 4 is used and it is sufficient to provide the path to the directory as ``storage_options.uri``; whereas for a version 1, the path to the db3 file needs to be provided.

By setting the version of the talker bag to 1, 2, 3 I found that:
- version 1, 2, and 3 need the path to the db3 file instead of the directory
- version 4 is fine with the directory
- version 1 produces a misleading error message ``SQL logic error``
- version 2 and 3 produce a bit more helpful messages ``RuntimeError: No storage could be initialized. Abort``

I don't think this needs to be tackled as part of this PR, but think that reader should transparently handle these and not require application logic. Currently, we use:

```python
        # TODO: workaround for older rosbag2 format versions
        if meta['version'] < 4:
            assert len(meta['relative_file_paths']) == 1
            uri = str(path.parent / meta['relative_file_paths'][0])
        else:
            uri = str(path.parent)
```

Code will soon be public. What do you think? Shall I open a separate issue for this?

Reply:
I agree that the reader should handle this case without the application having to worry about it. Creating a separate issue sounds good to me.

Reply:
The version number is actually dealt with in `rosbag2_cpp::SequentialReader::open()`, which is an existing infrastructure the Python API just calls. Do you get this behavior only with Python, or with C++ as well? How about the command line `ros2 bag play`?

This line I believe takes care of the problem you are describing https://github.com/ros2/rosbag2/blob/master/rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp#L95

The version is read into a field in the struct `rosbag2_storage::BagMetadata`, which can be retrieve at the reader level with `rosbag2_cpp::BaseReaderInterface::get_metadata()`. https://github.com/ros2/rosbag2/blob/master/rosbag2_storage/include/rosbag2_storage/bag_metadata.hpp#L36

There are actually some tests in `rosbag2_cpp` and `rosbag2_tests` that specifically test for different versions as well.

It would be worth checking whether you are getting the error separate from the Python API, to isolate whether the problem is in `rosbag2_py` or elsewhere in the existing infrastructure.

Reply:
@mabelzhang I currently don't have a C++ setup to test with. It is however trivial to test by simply adjusting the version set in ``talker/matadata.yaml``. My understanding is, that the additional keys of newer versions are ignored by older versions.

Reply:
Hmm I don't know all the nuances of the different versions to reproduce this. I believe the metadata files for different versions look a little different, for example, the qos profiles aren't there for the older versions. I can open a ticket now with limited information, but it would not be too helpful if I cannot describe the exact way to reproduce it, and whether it's a Python or C++ problem. I can look into this a little later and create a more informative issue.